src/llm/llmnoderesources.hpp

```diff
@@ -45,62 +47,83 @@
+class TextCallbackStreamer : public StreamerBase {
+protected:
+    Tokenizer m_tokenizer;
+    std::vector<int64_t> m_tokens_cache;
+    std::vector<int64_t> m_decoded_lengths;
+    size_t m_printed_len = 0;
 
 public:
-    TextStreamer(std::shared_ptr<ov::genai::Tokenizer> tokenizer) :
-        tokenizer(std::move(tokenizer)) {}
+    bool put(int64_t token) {
+        std::stringstream res;
+        m_tokens_cache.push_back(token);
+        std::string text = m_tokenizer.decode(m_tokens_cache);
+        m_decoded_lengths.push_back(text.length());
+
+        if (!text.empty() && '\n' == text.back() && text.size() > m_printed_len) {
+            // Flush the cache after the new line symbol
+            res << std::string_view{text.data() + m_printed_len, text.size() - m_printed_len};
+            m_tokens_cache.clear();
+            m_decoded_lengths.clear();
+            m_printed_len = 0;
+            return on_finalized_subword_callback(res.str());
+        }
 
-    std::optional<std::string> put(std::vector<int64_t>& tokens) {
-        for (auto token : tokens) {
-            tokenCache.push_back(token);
+        constexpr size_t delay_n_tokens = 3;
+        // In some cases adding the next token can shorten the text,
+        // e.g. when apostrophe removing regex had worked after adding new tokens.
+        // Printing several last tokens is delayed.
+        if (m_decoded_lengths.size() < delay_n_tokens) {
+            return on_finalized_subword_callback(res.str());
         }
-        std::string text = tokenizer->decode(tokenCache);
-        if (!text.empty() && '\n' == text.back() && text.size() > printLen) {
-            // The chunk is ready if the generated text ends with new line.
-            // Also, clear the cache.
-            std::string chunk = std::string{text.data() + printLen, text.size() - printLen};
-            SPDLOG_LOGGER_TRACE(llm_calculator_logger, "Generated tokens: {}", tokenCache);
-            tokenCache.clear();
-            printLen = 0;
-            return chunk;
-        } else if (!isValidUtf8(text)) {
-            return std::nullopt;
-        } else if (text.size() > printLen) {
-            // The chunk is ready if the new text in the cache contains space.
-            // The chunk is constructed from the new text, however only up to the last space character (including it)
-            // Does not clear the cache.
-            auto lastSpacePos = text.rfind(' ');
-            if (lastSpacePos == std::string::npos || lastSpacePos < printLen) {
-                return std::nullopt;
-            }
-            std::string chunk = std::string{text.data() + printLen, lastSpacePos - printLen + 1};
-            printLen = lastSpacePos + 1;
-            SPDLOG_LOGGER_TRACE(llm_calculator_logger, "Generated tokens: {}", tokenCache);
-            return chunk;
+        constexpr char replacement[] = "\xef\xbf\xbd";  // MSVC with /utf-8 fails to compile <UNK> directly with newline in string literal error.
```

Comment:
In this comment, GenAI contains special character instead of `<UNK>`. I had to replace it, so our pipeline passes.